File: platforms/ios/Dezel/DezelTests/Core/JavaScriptSetterTest.swift
import XCTest
import Dezel

class JavaScriptSetterTest: XCTestCase {

	private var context: JavaScriptContext = JavaScriptContext()

	override func setUp() {
		super.setUp()
		self.context.dispose()
		self.context = JavaScriptContext()
		self.context.setup()
	}

	func testCallback() {

		var didCallCallback = false
		var expectedCallee: JavaScriptValue?
		var expectedObject: JavaScriptValue?

		let obj = self.context.createEmptyObject()

		self.defineSetter({ (callback: JavaScriptSetterCallback) in
			expectedCallee = callback.callee
			expectedObject = callback.target
			didCallCallback = true
		}, target: obj).property("property", number: 10)

		XCTAssertNotNil(expectedCallee)
		XCTAssertNotNil(expectedObject)
		XCTAssertTrue(obj.equals(expectedObject!))

		expectedCallee?.dispose()
		expectedObject?.dispose()

		obj.dispose()

		XCTAssertTrue(didCallCallback)
	}

	func testValue() {

		func setter(callback: JavaScriptSetterCallback) {
			XCTAssertTrue(callback.value.isString)
		}

		let obj = self.defineSetter(setter, target: nil)

		let value = self.context.createString("")
		obj.property("property", value: value)
		value.dispose()

		obj.dispose()
	}

	func testValueAsString() {

		func setter(callback: JavaScriptSetterCallback) {
			XCTAssertEqual(callback.value.string, "")
		}

		let obj = self.defineSetter(setter, target: nil)
		obj.property("property", string: "")
		obj.dispose()
	}

	func testValueAsNumber() {

		func setter(callback: JavaScriptSetterCallback) {
			XCTAssertEqual(callback.value.number, 10)
		}

		let obj = self.defineSetter(setter, target: nil)
		obj.property("property", number: 10)
		obj.dispose()
	}

	func testValueAsBoolean() {

		func setter(callback: JavaScriptSetterCallback) {
			XCTAssertEqual(callback.value.boolean, true)
		}

		let obj = self.defineSetter(setter, target: nil)
		obj.property("property", boolean: true)
		obj.dispose()
	}

	//--------------------------------------------------------------------------
	// Private API
	//--------------------------------------------------------------------------

	internal func defineSetter(_ setter: @escaping JavaScriptSetterHandler, target: JavaScriptValue?) -> JavaScriptValue {

		var obj = target
		if (obj == nil) {
			obj = self.context.createEmptyObject()
		}

		obj!.defineProperty("property", value: nil, getter: nil, setter: setter, writable: false, enumerable: false, configurable: false)

		return obj!
	}

	internal func defineGetter(_ getter: @escaping JavaScriptGetterHandler, target: JavaScriptValue?) -> JavaScriptValue {

		var obj = target
		if (obj == nil) {
			obj = self.context.createEmptyObject()
		}

		obj!.defineProperty("property", value: nil, getter: getter, setter: nil, writable: false, enumerable: false, configurable: false)

		return obj!
	}
}

File: platforms/android/Dezel/dezel/src/androidTest/java/ca/logaritm/dezel/core/JavaScriptSetterTest.kt
package ca.logaritm.dezel.core

import org.junit.Assert
import org.junit.Before
import org.junit.Test

class JavaScriptSetterTest {

	private var context: JavaScriptContext = JavaScriptContext()

	@Before
	fun beforeTest() {
		this.context.dispose()
		this.context = JavaScriptContext()
	}

	@Test
	fun testCallback() {

		val obj = this.context.createEmptyObject()

		var didCallCallback = false
		var expectedCallee: JavaScriptValue? = null
		var expectedObject: JavaScriptValue? = null

		this.defineSetter(fun(callback: JavaScriptSetterCallback) {
			expectedCallee = callback.callee
			expectedObject = callback.target
			didCallCallback = true
		}, obj).property("property", 10.0)

		Assert.assertNotNull(expectedCallee)
		Assert.assertNotNull(expectedObject)
		Assert.assertTrue(obj.equals(expectedObject!!))

		expectedCallee?.dispose()
		expectedObject?.dispose()

		obj.dispose()

		Assert.assertTrue(didCallCallback)
	}

	@Test
	fun testValue() {

		val setter = fun(callback: JavaScriptSetterCallback) {
			Assert.assertTrue(callback.value.isString)
		}

		val obj = this.defineSetter(setter, null)

		val value = this.context.createString("")
		obj.property("property", value)
		value.dispose()

		obj.dispose()
	}

	@Test
	fun testValueAsString() {

		val setter = fun(callback: JavaScriptSetterCallback) {
			Assert.assertEquals(callback.value.string, "")
		}

		val obj = this.defineSetter(setter, null)
		obj.property("property", "")
		obj.dispose()
	}

	@Test
	fun testValueAsNumber() {

		val setter = fun(callback: JavaScriptSetterCallback) {
			Assert.assertEquals(callback.value.number, 10.0, 0.0)
		}

		val obj = this.defineSetter(setter, null)
		obj.property("property", 10.0)
		obj.dispose()
	}

	@Test
	fun testValueAsBoolean() {

		val setter = fun(callback: JavaScriptSetterCallback) {
			Assert.assertEquals(callback.value.boolean, true)
		}

		val obj = this.defineSetter(setter, null)
		obj.property("property", true)
		obj.dispose()
	}

	//--------------------------------------------------------------------------
	// Private API
	//--------------------------------------------------------------------------

	fun defineSetter(setter: JavaScriptSetterHandler, target: JavaScriptValue?): JavaScriptValue {

		var target = target
		if (target == null) {
			target = context.createEmptyObject()
		}

		target.defineProperty("property", null, null, setter, false, false, false)

		return target
	}

	fun defineGetter(getter: JavaScriptGetterHandler, target: JavaScriptValue?): JavaScriptValue {

		var target = target
		if (target == null) {
			target = context.createEmptyObject()
		}

		target.defineProperty("property", null, getter, null, false, false, false)

		return target
	}
}

